Add context menu to quickly filter bundles (resolves #241)

bregenspan opened the pull request:
As discussed in #241, adding a context menu that exposes options for each chunk can allow for an easier way to trigger certain operations, such as filtering to show only a single chunk, filtering to show all chunks, filtering down to parents of a chunk, or hiding a particular chunk.

This is a first stab at adding such a right-click/Ctrl-click triggered context menu for chunk options, with these options to start with:

 * Hide chunk
 * ~Show parent chunks~ (removed for now, see comments)
 * Filter to chunk
 * Show all chunks

One possible limitation of the current approach is that I went for a bespoke context menu based partly on the existing Tooltip component. To add features like a multi-level hierarchy in the context menu in future or more dynamic/context-specific options, it might make more sense to replace with an existing full-featured open-source component.

TODOs:
- [x] Bug: Clicking outside the context menu closes it. However, the same click event that closes the context menu can cause interaction with the treemap (e.g. a zoom-in on the chunk group that was clicked), which can feel a bit buggy. (Not sure if this is a blocker, but I do want to take another look at fixing).
- [x] Bug: menu position flickers/shows in upper left briefly upon close (see https://github.com/webpack-contrib/webpack-bundle-analyzer/pull/246#issuecomment-480481095)
- [x] Bug: Lack of handling for window resize while menu is open
- [x] Bug/missing feature: It is possible to hide all chunks until none appear and then the only way to get chunks showing again is to use the sidebar menu or refresh the page
- [x] When menu is already open, right click/ctrl-click on another chunk when menu is open should immediately open the menu for that chunk, instead of closing the menu and requiring another click to re-open (see https://github.com/webpack-contrib/webpack-bundle-analyzer/pull/246#discussion_r272918615)
- [ ] (last tiny one): support opening the menu via `event.metaKey` in addition to `event.ctrlKey` 

Possible future cleanup:
- Better accessibility story? But making the context menu itself screen reader-accessible is pretty clear - the really hard/meaningful part would be getting the treemap itself, and all interactions leading up to the opening of the context menu, to be meaningful! That'd be a pretty big and interesting challenge, probably involves some very thoughtful changes to the actual underlying visualization, so probably makes best sense to address separately.

![example screencast](https://user-images.githubusercontent.com/237410/51811354-97d23100-227a-11e9-947f-924d3dd9e8d4.gif)

bregenspan commented on client/components/ContextMenu.jsx:
```diff
@@ -0,0 +1,118 @@
+        <li className={itemClassName} onClick={this.handleClickHideChunk}>Hide chunk</li>
+        <li className={itemClassName} onClick={this.handleClickFilterToParents}>Show parent chunks</li>
+        <li className={itemClassName} onClick={this.handleClickFilterToChunk}>Hide all other chunks</li>
+        <hr/>
```
I might have made the laziest possible interpretation of @th0r's request in https://github.com/webpack-contrib/webpack-bundle-analyzer/issues/241#issuecomment-456334239 to add group separators here.

I could see either abstracting out the ContextMenu component so it can be passed a declarative config of options OR swapping out in favor of an existing open-source context menu, as there might be one that already has good UX thought put into it and that supports hierarchical menu options.

th0r commented:
> * Clicking outside the context menu closes it. However, the same click event that closes the context menu can cause interaction with the treemap

What about adding a capture click event on document and preventing its propagation?

```js
handleMenuOpen() {
  // ...
  document.addEventListener('click', event => {
    if (<event_target_is_outside_context_menu>) {
      event.preventDefault();
      event.stopPropagation();
    }
  }, true);
}
```

bregenspan commented:
@th0r that did it, thanks -- just pushed that fix up!

valscion commented:
Hi there @bregenspan, sorry that you've had to wait for so long with this PR.

I took this PR for a spin now and it feels to be working nicely. However, I found myself quite confused on the terminology, and I'm worried that people not so knowledgeable in webpack will have an even harder time to figure out what "parent chunk" means.

At the very least, we would need to document what "parent chunk" means in the readme. How would you describe this to a newcomer?

valscion commented:
There's also this strange flicker of the tooltip after closing it in Firefox at least:

![Kapture 2019-04-06 at 10 13 14](https://user-images.githubusercontent.com/482561/55666254-b5141480-5854-11e9-952f-4956295996ba.gif)

See how the tooltip moves to the top left when it is being closed?

bregenspan commented:
@valscion thank you for checking out this PR, and for good points both re: documentation and that visual bug! I just pushed up a first attempt at some basic documentation describing the context menu and the existing feature for showing/hiding bundles (sidebar), as I think this section could be a good anchor for describing the "parent chunk" concept, which is definitely the most complicated and unclear of the context menu options. I'm realizing that I can only describe what a "parent chunk" is in somewhat hazy terms or with reference to my specific project setup, so will look into that more and work to find the right definition. (Also checking on that visual bug!)

bregenspan commented:
@valscion I fixed the bug you noted, thanks for catching that! Also, related to the point you brought up re: "parent chunks" not being an obvious concept: I agree, and I think it could make the most sense to not include the option here and trim down the scope of the PR (just pushed a commit doing so, and updated PR description accordingly). Then I can add the parent chunk logic separately in a small, easier-to-review PR once I've written really solid explanation of what parent chunk actually means and why someone would want to filter to them. Does that make sense to you?

valscion commented:
Thanks! I agree with you that it is better to first ship this feature without the parent chunk option and later add it back to make the scope of this PR more manageable. Good call :+1:

The context menu bug is now indeed fixed :+1:. I also like how you handle resizing by closing the context menu — it's a nice easy way to get out of a weird situation.

How do you feel about it being possible to accidentally hide all the chunks with the context menu? I found myself a bit lost when I hid the last visible chunk. Should we only allow showing all the chunks when only one chunk is visible? Or should we have some sort of a large view showing text "all chunks hidden, [click here to show all chunks]()" when all chunks are hidden?

![Kapture 2019-04-07 at 10 51 29](https://user-images.githubusercontent.com/482561/55680525-2f12cf00-5923-11e9-833e-bf9405ad5ee3.gif)

valscion commented on README.md:
```diff
@@ -148,6 +148,22 @@ as Uglify, then this value will reflect the minified size of your code.
 
 This is the size of running the parsed bundles/modules through gzip compression.
 
+<h2 align="center">Selecting Which Chunks to Display</h2>
```
This is a nice piece of documentation :+1:

bregenspan commented:
> How do you feel about it being possible to accidentally hide all the chunks with the context menu? I found myself a bit lost when I hid the last visible chunk. Should we only allow showing all the chunks when only one chunk is visible? Or should we have some sort of a large view showing text "all chunks hidden, click here to show all chunks" when all chunks are hidden?

Good point! I was thinking of the first of these ideas - disabling the "hide chunk" option when only one is visible. That would also be a good opportunity to refactor out the menu items a bit. I could go for either approach, but approach of disabling menu options has the benefit that we can also disable the "show all chunks" option when all chunks are already showing, so I lean towards that one. I can add that to this PR.

bregenspan commented:
@valscion I pushed up one of those two possible solutions, what do you think?

These are the three current states:

#### One chunk is selected
<img width="236" alt="1-chunk-selected" src="https://user-images.githubusercontent.com/237410/55688069-5b8a1380-5942-11e9-8508-63cab1b720d8.png">

#### All chunks are selected
<img width="214" alt="all-chunks-selected" src="https://user-images.githubusercontent.com/237410/55688070-5b8a1380-5942-11e9-9086-52bc3e0e5e92.png">

#### More than 1 chunk, but fewer than all chunks, selected
<img width="216" alt="more-than-1-less-than-all-selected" src="https://user-images.githubusercontent.com/237410/55688071-5b8a1380-5942-11e9-9271-c0924331bf3d.png">

bregenspan commented on client/components/ContextMenu.jsx:
```diff
@@ -0,0 +1,114 @@
+
+export default class ContextMenu extends PureComponent {
+  componentDidMount() {
+    this.boundingRect = this.node.getBoundingClientRect();
```
Right now the contents of the menu are static, so this optimization makes sense. It'll be necessary to update this strategy if the menu is made dynamic enough that its sizing can change.

bregenspan commented on client/components/ContextMenu.jsx:
```diff
@@ -0,0 +1,114 @@
+    const multipleChunksSelected = store.selectedChunks.length > 1;
+    return (
+      <ul ref={this.saveNode} className={containerClassName} style={this.getStyle()}>
+        <ContextMenuItem disabled={!multipleChunksSelected}
```
This is a bit of a digression, but I think it would be really great if this markup generated here was more screen reader accessible. The main challenge seems to not be in the webpack-bundle-analyzer-specific controls, but in making the whole underlying visualization accessible, such that a user could even get to a point where the controls are meaningful to use (curious if anyone does know of a common screen reader that's able to generate a meaningful result for the treemap content itself, it seems like a hard challenge).

valscion commented:
Seems good to me! I'll let @th0r take a look if he's interested. To me this seems ready and could be released in a minor release.

bregenspan commented on client/components/ContextMenuItem.css:
```diff
@@ -0,0 +1,24 @@
+.item {
+  -webkit-user-select: none;
```
Good call, just opened an issue here: https://github.com/webpack-contrib/webpack-bundle-analyzer/issues/255

bregenspan commented on client/components/Treemap.jsx:
```diff
@@ -91,6 +91,10 @@ export default class Treemap extends Component {
       },
       onGroupClick(event) {
         preventDefault(event);
+        if ((event.ctrlKey || event.secondary) && props.onGroupSecondaryClick) {
```
Re: `event.metaKey`, Chrome and Firefox in OSX trigger the native context menu only when `e.ctrlKey === true`, not `e.metaKey === true`. But I should test in more browsers/OSes again and also don't see a downside to adding an extra trigger key, it could make sense to add that either way.

Also I can document that this is a FoamTree event because that's less-than-obvious -- would you be averse to using a JsDoc-style comment for that?

th0r replied:
> would you be averse to using a JsDoc-style comment for that

That's up to you, but I think just a simple one-liner will work as well.

bregenspan commented on client/components/ContextMenu.jsx:
```diff
@@ -0,0 +1,114 @@
+
+  handleDocumentMousedown = (e) => {
+    if (elementIsOutside(e.target, this.node)) {
+      e.preventDefault();
```
Fixed! There is one remaining inconsistency, which is that the menu can be opened via a Foamtree `event.secondary === true` event, which can include touch gestures. In such a case, the menu will still need two gestures to close and then re-open. Unfortunately, it's hard to get 1:1 parity between the handling of the Foamtree event, which doesn't expose its underlying DOM event, and this actual DOM event, and I didn't want to resort to using undocumented Foamtree methods to try to get around this.

th0r replied:
I think we can live with that.

bregenspan commented:
@th0r I just looked a bit more into whether to add `metaKey` as a hotkey - is there a particular OS/browser combination where you think it'd be most useful?

Opening context menus via a secondary click triggered by Ctrl feels a bit like a forced fit on Windows, but I think Windows users are more likely to use an actual hardware right click, or are used to the special behavior or Menu key immediately opening an app-specific context menu (which I didn't see working great for this context-specific, not app-wide, case).

Besides that question, I think this might be ready now - all feedback should be addressed. Thanks again @th0r and @valscion for all your time reviewing this!

th0r commented:
> I just looked a bit more into whether to add `metaKey` as a hotkey

Ok, lets not do it.

th0r commented:
Awesome! Thanks a lot!

th0r commented:
Released in v3.3.0
